### PR TITLE
28-feature-internationalization

### PR DESCRIPTION
**Pull request by Segelzwerg**

### CODE REVIEW

File: invoice/models.py [hunk 1]
```diff
@@ -13,20 +13,25 @@
     DateField, UniqueConstraint, OneToOneField, Q, F, TextChoices
 from django.db.models.constraints import CheckConstraint
 from django.db.models.fields import DecimalField
+from django.utils.translation import gettext as _
```
Comment (iTitus) [inline invoice/models.py:16]: shouldnt this be gettext_lazy (according to the django docs)?

Comment (iTitus) [reply invoice/models.py:16]: Also applies to all other "static" translations

File: invoice/pdf_generator.py [hunk 2]
```diff
@@ -90,17 +91,17 @@ def render_lines_left_right(x, y, lines, line_offset=0, line_height=16):
 
     # Totals
     render_lines_left_right(-(A4_WIDTH - x_left), table_y_end, [
-        ["Net Total: ", invoice.net_total_string],
-        ["Total: ", invoice.total_string]
+        [f'{gettext("Net Total")}: ', invoice.net_total_string],
```
Comment (iTitus) [inline invoice/pdf_generator.py:94]: these should use formatting instead of simple concatenation

**Review by iTitus — CHANGES_REQUESTED** [review]
It seems to be pretty standard for Django to use the full english sentence with punctuation and spaces as the message id.
I don't like that. Can we use a slug-like key instead?

File: invoice/pdf_generator.py (hunk 2 above)
Comment (Segelzwerg) [reply invoice/pdf_generator.py:94]: sorry, what do you mean by that?

**Comment by Segelzwerg** [comment]
> It seems to be pretty standard for Django to use the full english sentence with punctuation and spaces as the message id. I don't like that. Can we use a slug-like key instead?

I don't see the advantage of it. As translation the context is often relevant for the translation. So, having to translate a whole sentence it does make sense to put everything in.

**Comment by iTitus** [comment]
> I don't see the advantage of it. As translation the context is often relevant for the translation. So, having to translate a whole sentence it does make sense to put everything in.

It stops mistakes. It's very easy to have a whitespace error or a wrong non-letter character that doesnt match.
So I think `"Vendor successfully created!"` should be `vendor.create.success` instead.

File: invoice/pdf_generator.py (hunk 2 above)
Comment (iTitus) [reply invoice/pdf_generator.py:94]: you are concatenating the number and the translation of "Net Total ".
It should instead use some kind of formatting. I dont know the API, might be `gettext_format("Net Total", amount=invoice.net_total_string)` or something.

Comment (Segelzwerg) [reply invoice/pdf_generator.py:94]: Like this? Will changes others then.

**Comment by Segelzwerg** [comment]
> > I don't see the advantage of it. As translation the context is often relevant for the translation. So, having to translate a whole sentence it does make sense to put everything in.
> 
> It stops mistakes. It's obvious now what needs to be translated and what is just literal strings.
> 
> It's very easy to have a whitespace error or a wrong non-letter character that doesnt match. So I think `"Vendor successfully created!"` should be `vendor.create.success` instead.

But this is who django works and not how we can choose it, right? I think we have no influence on the message.

File: invoice/pdf_generator.py (hunk 2 above)
Comment (Segelzwerg) [reply invoice/pdf_generator.py]: Formatting with one string does break the code as it expects the row to have two elements.

Comment (iTitus) [reply invoice/pdf_generator.py:94]: oh yeah, i guess this special case needs to stay separate

**Comment by iTitus** [comment]
> But this is how django works and not how we can choose it, right? I think we have no influence on the message.

Yeah django works like that.
But nothing stops us from adding another .po file for english/default localization right?
Would that work?

File: invoice/pdf_generator.py (hunk 2 above)
Comment (iTitus) [reply invoice/pdf_generator.py:94]: But maybe some other cases that combine text with numbers or text with text. Those should be formatted together so a translator can move the number freely in the text snippet and maybe even apply plural rules

Comment (Segelzwerg) [reply invoice/pdf_generator.py:94]: Got it. Will do

**Comment by Segelzwerg** [comment]
> > But this is how django works and not how we can choose it, right? I think we have no influence on the message.
> 
> Yeah django works like that. But nothing stops us from adding another .po file for english/default localization right? Would that work?

Puh, I don't know, but why not rely on the automated process instead of doing it manually?

File: invoice/pdf_generator.py (hunk 2 above)
Comment (Segelzwerg) [reply invoice/pdf_generator.py:94]: I checked there are no sentence like structures in the pdf gen that require it. And we would end up with `gettext` formats within an f-string.
I prefer to leave it as such.

**Comment by iTitus** [comment]
> > > But this is how django works and not how we can choose it, right? I think we have no influence on the message.
> > 
> > 
> > Yeah django works like that. But nothing stops us from adding another .po file for english/default localization right? Would that work?
> 
> Puh, I don't know, but why not rely on the automated process instead of doing it manually?

What happens if we have to do formatting in English, or pluralisation?
Will the formatting/plural directives be part of the MSGID?
I really don't like this

**Comment by Segelzwerg** [comment]
> > > > But this is how django works and not how we can choose it, right? I think we have no influence on the message.
> > > 
> > > 
> > > Yeah django works like that. But nothing stops us from adding another .po file for english/default localization right? Would that work?
> > 
> > 
> > Puh, I don't know, but why not rely on the automated process instead of doing it manually?
> 
> What happens if we have to do formatting in English, or pluralisation? Will the formatting/plural directives be part of the MSGID? I really don't like this

Yes pluralization is supported. And according to this the msgid cannot be changed: https://stackoverflow.com/a/75415383/421615

**Comment by Segelzwerg** [comment]
> > > > > But this is how django works and not how we can choose it, right? I think we have no influence on the message.
> > > > 
> > > > 
> > > > Yeah django works like that. But nothing stops us from adding another .po file for english/default localization right? Would that work?
> > > 
> > > 
> > > Puh, I don't know, but why not rely on the automated process instead of doing it manually?
> > 
> > 
> > What happens if we have to do formatting in English, or pluralisation? Will the formatting/plural directives be part of the MSGID? I really don't like this
> 
> Yes pluralization is supported. And according to this the msgid cannot be changed: https://stackoverflow.com/a/75415383/421615

They documentation says, never change the msgid: https://docs.djangoproject.com/en/5.1/topics/i18n/translation/#message-files

**Comment by iTitus** [comment]
> They documentation says, never change the msgid: https://docs.djangoproject.com/en/5.1/topics/i18n/translation/#message-files

Yep that needs to happen.
But what if you want to change the string after? Do you still change the MSGID or do you add a .po for english?
And after you changed it you might have `MSGID="Deleted sucesfuli." MSGSTR="Deleted successfully!".
Extreme example, I know - but still worth to think about.

**Comment by Segelzwerg** [comment]
> > They documentation says, never change the msgid: https://docs.djangoproject.com/en/5.1/topics/i18n/translation/#message-files
> 
> Yep that needs to happen. But what if you want to change the string after? Do you still change the MSGID or do you add a .po for english? And after you changed it you might have `MSGID="Deleted sucesfuli." MSGSTR="Deleted successfully!"`. Extreme example, I know - but still worth to think about.

Oh I see your point. Django is actually smart and puts the old ID as comment on top, so you can find your changes easily.

**Comment by Segelzwerg** [comment]
@iTitus I merged main and it is ready for review.

**Review by iTitus — APPROVED** [review]

**Pull request merged**